### PR TITLE
[FIX] mail: impossible to update image of user on large database

### PR DESCRIPTION
**Pull request by fmdl**
Description of the issue/feature this PR addresses:
On large database with million of messages made by one user.
When you try to update the image of user, it take lot of time (CPU time limit)

I don't know if it is an issue of ORM, but ORM try to read mail_ids of messages (3 million in my case) here : https://github.com/odoo/odoo/blob/14.0/odoo/models.py#L5843

@rco-odoo @xmo-odoo @tde-banana-odoo 

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by fmdl** [comment]
OPW #2524520

**Comment by xmo-odoo** [comment]
That's a fun one. After filling a local db wih 10 million messages from admin, it actually blows up with a memory error.

File: addons/mail/models/mail_message.py [hunk 1]
```diff
@@ -141,6 +141,11 @@ def default_get(self, fields):
     mail_ids = fields.One2many('mail.mail', 'mail_message_id', string='Mails', groups="base.group_system")
     canned_response_ids = fields.One2many('mail.shortcode', 'message_ids', string="Canned Responses", store=False)
 
+    @api.depends('author_id')
```
Comment (xmo-odoo) [inline addons/mail/models/mail_message.py:144]: might as well remove the `depends` entirely. I also don't think `compute_sudo` matters any.

But I pinged @rco-odoo to see if it woudln't be possible to fix the underlying concern, namely that when we walk up from a dependency we load all records at every level, including the leaf, which can lead to extreme amplification when the root is a very commonly linked record, and the leaf is a very *common* record (e.g. mail.message for which millions isn't an extreme number, or moves).

**Review by rco-odoo — APPROVED** [review]
Thanks!
@robodoo r+

**Comment by tde-banana-odoo** [comment]
Can't the "depends" be automatically guessed based on related parameter ? I suppose we would have the same kind of issue on other models ...

**Comment by fmdl** [comment]
@tde-banana-odoo the depends is automatically guessed in this case `author_avatar = fields.Binary(related='author_id.image_128')` it should be : `depends=['author_id.image_128'])`

But `depends=['author_id.image_128'])` generate a huge trigger, that why we have reduce the trigger to `depends=['author_id']`

**Comment by tde-banana-odoo** [comment]
Indeed, I did not really spot that difference. It is a matter of reducing the depends. For non stored fields I suppose we have few real use case of having to depends on the final field, only an onchange-like on the m2o is necessary.

**Comment by xmo-odoo** [comment]
@robodoo rebase-ff

> Indeed, I did not really spot that difference. It is a matter of reducing the depends. For non stored fields I suppose we have few real use case of having to depends on the final field, only an onchange-like on the m2o is necessary.

Yeah in this case we probably could do with a completely empty depends as the odds that we would update a message's partner at all let alone in a transaction where we have message avatars loaded, and access those avatars afterwards, seems basically nil.

**Comment by fmdl** [comment]
@tde-banana-odoo @xmo-odoo @rco-odoo this issue exist on other place, some exemple:
**Partner**
https://github.com/odoo/odoo/blob/14.0/addons/calendar/models/calendar_attendee.py#L35
https://github.com/odoo/odoo/blob/14.0/addons/project/models/project.py#L813
https://github.com/odoo/odoo/blob/14.0/addons/project/models/project.py#L607
https://github.com/odoo/odoo/blob/14.0/addons/purchase/models/purchase.py#L135
https://github.com/odoo/odoo/blob/14.0/addons/website_sale/models/account_move.py#L10
**Product**
https://github.com/odoo/odoo/blob/14.0/addons/event_sale/models/sale_order.py#L61
https://github.com/odoo/odoo/blob/14.0/addons/lunch/models/lunch_order.py#L14
https://github.com/odoo/odoo/blob/14.0/addons/mrp/models/mrp_production.py#L98
https://github.com/odoo/odoo/blob/14.0/addons/point_of_sale/models/pos_order.py#L753
https://github.com/odoo/odoo/blob/14.0/addons/product_expiry/models/production_lot.py#L11
https://github.com/odoo/odoo/blob/14.0/addons/purchase/models/purchase.py#L782
...

In large database if the depends is not restricted, the database can run slowly.
For the next version maybe make a check on all related fields.

Or if the field is related and non stored, the depends is automatically only the first field.

cc @Yenthe666 @alexis-via @pedrobaeza

**Comment by xmo-odoo** [comment]
> Or if the field is related and non stored, the depends is automatically only the first field.

That's absolutely not correct so no. Here we're trading strict correctness away because the situation where it could occur has essentially no chance of happening. That's not a valid assumption in general.

**Comment by Yenthe666** [comment]
@fmdl why not against V12/V13 and then forward port? Now it is only fixed in V14 :) 
That being said: I also have a similar issue on a V13 db with 5M+ messages where we can't get the messages open anymore hah.

**Comment by rco-odoo** [comment]
@Yenthe666 it may be backported to v13, but not to v12.
That's because the possibility to "force" `depends` only exists since Odoo 13.0.

**Comment by Yenthe666** [comment]
@fmdl do you wish to make one against V13 or should I make one? Let's do it there too :)

**Comment by xmo-odoo** [comment]
> @Yenthe666 it may be backported to v13, but not to v12.
> That's because the possibility to "force" `depends` only exists since Odoo 13.0.

Could be backported to 12.0 by converting the related to a normal compute tho.

**Pull request closed without merging**